felt/src/bigint_felt.rs

```diff
@@ -170,7 +170,22 @@ impl FeltOps for FeltBigInt {
     }
 
     fn sqrt(&self) -> Self {
-        FeltBigInt(self.0.sqrt())
+        //FeltBigInt(self.0.sqrt())
+        // Based on Tonelli-Shanks' algorithm for finding square roots
+        // and sympy's library implementation
+        let max_felt = FeltBigInt::max_value();
+        let trailing_prime = FeltBigInt::max_value() >> 48;
+        let a = self.pow(&trailing_prime);
+        let d = (&FeltBigInt::new(3_i32)).pow(&trailing_prime);
+        let mut m = FeltBigInt::zero();
+        for i in 0..48 {
+            let mut adm = &a * &(&d).pow(&m);
+            adm = adm.pow(1_u64 << (47 - i));
+            if &adm == &max_felt {
+                m += 1 << i;
+            }
+        }
```

Comment:
I think this might help:
```suggestion
        let d = (&FeltBigInt::new(3_i32)).pow(&trailing_prime);
        let d_pow = d.clone();
        let mut adm = a.clone();
        for i in 0..48 {
            adm = adm.pow(1_u64 << (47 - i));
            if &adm == &max_felt {
                d_pow *= d_pow * d;
            }
        }
```